### ISSUE TITLE
Add a `bcachefs` storage driver

### ISSUE THREAD

**Issue by stgraber** [lxc/incus#213]
Linux 6.7 should finally have initial support for `bcachefs` so we should be able to add a storage driver for it, roughly based on the btrfs or zfs one (whichever ends up closest).

@RAOF showed some initial implementation of that for LXD a while back, so that shouldn't take too much effort to get that in place and set up some testing for it.

Marking as blocked until Linux 6.7 is released.

**Comment by stgraber** [comment]
@RAOF did you push your LXD branch anywhere public?

**Comment by RAOF** [comment]
I did not. I should dust that off and put it somewhere useful!

**Comment by FlyingWombat** [comment]
bcachefs is now in mainline kernel v6.7 , released 2024-01-07.
It is supported by at least Archlinux and Nixos as of the v6.7 kernel release.
Bcachefs is feature equivalent with Btrfs and Zfs, as far as LXD is concerned -- with the exception of a dedicated subvolume send/receive mechanism.

**Comment by CharanSriramUni** [comment]
Hey! Was also interested in approaching this issue. 

Is there any additional info available/any good resources to look at for reference?

**Comment by stgraber** [comment]
That one is likely to be a rather large one to work on.

I've not actually done much research into bcachefs as it's such a new feature (Linux 6.7) but I'd probably start any work on this by first familiarizing myself with the filesystem and its tools, then would probably pick the existing btrfs storage driver (`internal/storage/drivers/driver_btrfs*`) as a rough template for the work, looking for equivalent commands and logic under bcachefs.

**Comment by ktran1403** [comment]
Hello, my group is interested in working on this issue. Is it possible for us to be assigned to this? Thanks.

**Comment by stgraber** [comment]
@ktran1403 given my previous comment, I suspect it would be very unlikely that your group would have sufficient time to bring this to completion. I estimate this work at a couple of weeks of full time work by someone already familiar with both bcachefs and Incus.

**Comment by ktran1403** [comment]
@stgraber I see. Thanks for the heads-up.

**Comment by milaiwi** [comment]
Since no one else is working on this, would it be possible to break this into separate tasks/issues so we could tackle a somewhat difficult aspect of this project? We've familiarized ourselves with the codebase already given our other tasks but would love to contribute more to this project. If that's possible let us know, if not we understand!

**Comment by stgraber** [comment]
@milaiwi the first part would be getting a system with a functional bcachefs setup, so that means a 6.7 or 6.8 (probably best) Linux kernel and the bcachefs tools setup.

I'd probably then just play a bit with bcachefs, familiarize yourself with the concepts, start creating subvolumes, snapshots, setting up quotas, ... At which point you should have a good idea of what commands need to be run to deal with it, which will then have to be added to the new driver.

**Comment by TinkeringWithUS** [comment]
Can you give an example of how we would test out the new driver and whether if Incus is detecting Bcachefs as a valid filesystem?

**Comment by stgraber** [comment]
Once you've added a minimal driver for it, restarting incus should then have the driver's load function be called and if that succeeds, it will show up in the `incus info` data.

**Comment by milaiwi** [comment]
Yeah, we gave a stab at this but this looks like something that would need people who are more comfortable with bcachefs, as you mentioned. From what I've read bcachefs is pretty different from btrfs so we would have had to do a lot more designing.

**Comment by stgraber** [comment]
Yeah, it shouldn't be too too different from btrfs but it's definitely a fully new storage driver with a different way to create the storage and different commands for most of the other interactions. I also expect things like migration and quotas to work pretty differently.

**Comment by ktran1403** [comment]
Hello, while this issue is no longer possible for us to finish in time for class, I still find it rather interesting. Would it be possible for me to continue working on this? So far, I've familiarized myself with bcachefs and managed to get Incus to recognize bcachefs as well as create volumes and snapshots, but there's still a lot of work to be done.

**Comment by stgraber** [comment]
@ktran1403 for sure! I've assigned it to you so anyone else looking at our open issues is aware that you've started some work on this.

If you end up not having the time to work on this, just drop whatever notes you have in this issue and someone else can take it over.

**Comment by thomasdba** [comment]
any updates？

**Comment by stgraber** [comment]
Unassigning due to inactivity so others can have a go at it.

**Comment by stgraber** [comment]
@Aryan470 I'd recommend against taking this one. I actually have just labeled it as a "Maybe" now as bcachefs has been going through quite a few controversies recently which don't fill me with confidence about his future :)

**Comment by Aryan470** [comment]
@stgraber Okay, thanks for the heads up!

**Comment by victoitor** [comment]
Since I have been following bcachefs updates from time to time, I'd like to add my 2 cents on this. 

Indeed bcachefs has been going through some controversies recently which is not great to see. These controversies have mostly involved the project leader of bcachefs on the transition to the linux kernel. At least from my part, I still have hope this will eventually be resolved and these issues will settle down, and if this happens, maybe it might make sense to come back to this later.

I mostly wanted to say that, even if this is indeed the case, this issue should he handled with care. Filesystems can be awesome but handle very important data and bcachefs is still flagged as experimental. If incus would implement a driver for it, people could see it as a sign that it's "good enough" for use in production. This could be bad for the incus project itself as complaints on bcachefs usage may flood incus forums (as complaints on other storage drivers do as well).

For those who are not aware, in the current state of bcachefs development, the main delevolper is mostly solving a bunch of bugs, which are expected to be found in such a huge codebase and with the new user base of being included in the kernel. I for one would only try after it has lost the experimental flag for at least one or two years to get some maturity in its tools. Maybe incus could wait a little while as well. New things are nice, but filesystems need some extra maturity independent of these controversies.

**Comment by symgryph** [comment]
Linus isn't too partial to BcacheFS as of this moment. Some kerfuffle with the devs or something.

**Comment by rsyring** [comment]
FWIW, here's a bcachefs updating [on moving to DKMS ](https://lore.kernel.org/linux-bcachefs/5tle2holwzznjhuqvpplyemh5ykwsjamd4cypf2wahyyyu2mca@263qi2bnoeiq/T/#m5e1493fed7618ffaa49e448021999af626608a5c) for distribution.

And a [HN discussion](https://news.ycombinator.com/item?id=45423004) on bcachefs getting removed from the kernel.

**Comment by FlyingWombat** [comment]
Bcachefs getting removed from the kernel sounds bad, but it isn't a problem. It is ultimately a political issue, not a technical one. ZFS isn't in the kernel.

The move to DKMS has some benefits. For one, the driver is now more portable -- which will make it easier to port to other platforms.

So long as there are people using Bcachefs, and someone willing to contribute this driver, this issue can move forward.

**Comment by bensmrs** [comment]
Given the recent developments, I’d say we can finally close this issue :)

**Comment by stgraber** [comment]
Yeah, we can always revisit later but I think we'd want to see major distributions step up support on it first before we invest time adding and maintaining a driver.

Things were looking a lot better for that to happen back when the issue first got opened than it does now.

**Comment by perdakovich** [comment]
zfs is not in the kernel as well but it works, the issue of bcachefs is only because it's not as popular, correct?

**Comment by stgraber** [comment]
Correct, per my previous comment, I never said that we're closing this because bcachefs isn't mainline, we're closing it because the high interest and demand we once has all but vanished.

Once we see Linux distributions include it directly or make installing it from their package archive work in a way that's similar to ZFS, then we can look into it again.

I just had a quick look on the Debian side of things and the tools are only in experimental and I'm not seeing any DKMS type package for the filesystem itself.

**Comment by Strat00s** [comment]
@stgraber I guess the time has almost come? They removed the experimental label in June. On Arch and NixOS, you need to install `bcachefs-tools` and you are done. On Debian, you need to add the bcachefs repo, install `bcachefs-tools` and you're also done.
